File: parqueadero-service/src/main/java/com/cicloparqueadero/parqueadero/repository/BicicletaRepository.java
package com.cicloparqueadero.parqueadero.repository;

import com.cicloparqueadero.parqueadero.entity.Bicicleta;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface BicicletaRepository extends JpaRepository<Bicicleta, Long> {

    long countByEstado(Bicicleta.Estado estado);

    @Query("""
            SELECT b FROM Bicicleta b
            WHERE (:estado IS NULL OR b.estado = :estado)
              AND (:search IS NULL OR :search = ''
                   OR LOWER(b.propietario)    LIKE LOWER(CONCAT('%', :search, '%'))
                   OR LOWER(b.documento)      LIKE LOWER(CONCAT('%', :search, '%'))
                   OR LOWER(b.tipoBicicleta)  LIKE LOWER(CONCAT('%', :search, '%'))
                   OR LOWER(b.color)          LIKE LOWER(CONCAT('%', :search, '%')))
            ORDER BY b.createdAt DESC
            """)
    List<Bicicleta> search(@Param("estado") Bicicleta.Estado estado,
                           @Param("search") String search);
}

File: parqueadero-service/src/repositories/bicicleta.repository.js
const { Op } = require('sequelize');
const Bicicleta = require('../models/bicicleta.model');

/**
 * Capa de acceso a datos para bicicletas.
 */
class BicicletaRepository {
  async create(data) {
    return Bicicleta.create(data);
  }

  async findAll(filter = {}) {
    const where = {};
    if (filter.estado) where.estado = filter.estado;
    if (filter.search) {
      where[Op.or] = [
        { propietario: { [Op.like]: `%${filter.search}%` } },
        { documento: { [Op.like]: `%${filter.search}%` } },
        { tipo_bicicleta: { [Op.like]: `%${filter.search}%` } },
        { color: { [Op.like]: `%${filter.search}%` } },
      ];
    }
    return Bicicleta.findAll({ where, order: [['createdAt', 'DESC']] });
  }

  async findById(id) {
    return Bicicleta.findByPk(id);
  }

  async update(id, data) {
    const bicicleta = await Bicicleta.findByPk(id);
    if (!bicicleta) return null;
    await bicicleta.update(data);
    return bicicleta;
  }

  async delete(id) {
    const bicicleta = await Bicicleta.findByPk(id);
    if (!bicicleta) return false;
    await bicicleta.destroy();
    return true;
  }

  async countByEstado(estado) {
    return Bicicleta.count({ where: { estado } });
  }

  async count() {
    return Bicicleta.count();
  }
}

module.exports = new BicicletaRepository();
